src/backend/distributed/commands/multi_copy.c

```diff
@@ -0,0 +1,1177 @@
+	copyState = BeginCopyFrom(rel, copyStatement->filename,
+							  copyStatement->is_program,
+							  copyStatement->attlist,
+							  copyStatement->options);
+
+	/* set up callback to identify error line number */
+	errorCallback.callback = CopyFromErrorCallback;
+	errorCallback.arg = (void *) copyState;
+	errorCallback.previous = error_context_stack;
+	error_context_stack = &errorCallback;
+
+	executorState = CreateExecutorState();
+	executorTupleContext = GetPerTupleMemoryContext(executorState);
+	executorExpressionContext = GetPerTupleExprContext(executorState);
+
+	copyOutState = (CopyOutState) palloc0(sizeof(CopyOutStateData));
+	copyOutState->binary = true;
+	copyOutState->fe_msgbuf = makeStringInfo();
+	copyOutState->rowcontext = executorTupleContext;
+
+	columnOutputFunctions = ColumnOutputFunctions(tupleDescriptor, copyOutState->binary);
+
+	/* we use a PG_TRY block to roll back on errors (e.g. in NextCopyFrom) */
+	PG_TRY();
+	{
+		while (true)
+		{
+			bool nextRowFound = false;
+			Datum partitionColumnValue = 0;
+			ShardInterval *shardInterval = NULL;
+			int64 shardId = 0;
+			bool found = false;
+			MemoryContext oldContext = NULL;
+
+			ResetPerTupleExprContext(executorState);
+
+			oldContext = MemoryContextSwitchTo(executorTupleContext);
+
+			/* parse a row from the input */
+			nextRowFound = NextCopyFrom(copyState, executorExpressionContext,
+										columnValues,columnNulls, NULL);
+
+			MemoryContextSwitchTo(oldContext);
+
+			if (!nextRowFound)
+			{
+				break;
+			}
+
+			CHECK_FOR_INTERRUPTS();
+
+			/* find the partition column value */
+
+			if (columnNulls[partitionColumn->varattno - 1])
+			{
+				ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+								errmsg("cannot copy row with NULL value "
+									   "in partition column")));
+			}
+
+			partitionColumnValue = columnValues[partitionColumn->varattno - 1];
+
+			/* find the shard interval and id for the partition column value */
+			shardInterval = FindShardInterval(partitionColumnValue, shardIntervalCache,
+											  shardCount, partitionMethod,
+											  compareFunction, hashFunction,
+											  useBinarySearch);
+			if (shardInterval == NULL)
+			{
+				ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+								errmsg("no shard for partition column value")));
+			}
+
+			shardId = shardInterval->shardId;
+
+			/* find the connections to the shard placements */
+			shardConnections = (ShardConnections *) hash_search(shardConnectionHash,
+																&shardInterval->shardId,
+																HASH_ENTER,
+																&found);
+			if (!found)
+			{
+				/* intialize COPY transactions on shard placements */
+				shardConnections->shardId = shardId;
+				shardConnections->connectionList = NIL;
+
+				OpenCopyTransactions(copyStatement, shardConnections, shardId);
+
+				BuildCopyBinaryHeaders(copyOutState);
+				SendCopyDataToPlacements(copyOutState->fe_msgbuf, shardConnections);
```

Comment:
BuildCopy functions first clear the string buffer, then fill it. SendCopyDataToPlacements only uses the string. Do you think that we need to clear string buffer at the end of SendCopyDataToPlacements()?